File: algorithms/trivial_structs/stack.py
# Stack
# Time Complexity:
# push():       O(1) amortized
# pop():        O(1)
# peek():       O(1)
# is_empty():   O(1)
# len():        O(1)
#
# Space Complexity:
# Memory: O(n)

class Stack:
    def __init__(self):
        self.data = []

    def push(self, data):
        self.data.append(data)

    def pop(self):
        if not self.data:
            raise IndexError("pop from empty stack")

        return self.data.pop()

    def peek(self):
        if not self.data:
            raise IndexError("peek from empty stack")

        return self.data[-1]

    def is_empty(self):
        return len(self.data) == 0

    def __len__(self):
        return len(self.data)

File: cpp_algorithms/trivial_structs/stack.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

// Stack
//
// Time Complexity:
// push():      O(1) amortized
// pop():       O(1)
// peek():      O(1)
// is_empty():  O(1)
// size():      O(1)
//
// Space Complexity:
// O(n)

class Stack {
private:
    vector<int> data;

public:
    void push(int value) {
        data.push_back(value);
    }

    int pop() {
        if (data.empty()) {
            throw out_of_range("pop from empty stack");
        }

        int value = data.back();
        data.pop_back();

        return value;
    }

    int peek() const {
        if (data.empty()) {
            throw out_of_range("peek from empty stack");
        }

        return data.back();
    }

    bool is_empty() const {
        return data.empty();
    }

    int size() const {
        return data.size();
    }
};
